llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -4201,5 +4201,23 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
     }
   }
 
+  // select (trunc nuw X to i1), X, Y --> select (trunc nuw X to i1), 1, Y
+  // select (trunc nuw X to i1), Y, X --> select (trunc nuw X to i1), Y, 0
+  // select (trunc nsw X to i1), X, Y --> select (trunc nsw X to i1), -1, Y
+  // select (trunc nsw X to i1), Y, X --> select (trunc nsw X to i1), Y, 0
+  Value *Trunc;
+  if (match(CondVal, m_NUWTrunc(m_Value(Trunc)))) {
+    if (match(TrueVal, m_Specific(Trunc)))
+      return replaceOperand(SI, 1, ConstantInt::get(TrueVal->getType(), 1));
+    if (match(FalseVal, m_Specific(Trunc)))
+      return replaceOperand(SI, 2, ConstantInt::get(TrueVal->getType(), 0));
+  }
+  if (match(CondVal, m_NSWTrunc(m_Value(Trunc)))) {
+    if (match(TrueVal, m_Specific(Trunc)))
+      return replaceOperand(SI, 1, ConstantInt::get(TrueVal->getType(), -1));
```

Comment:
```suggestion
      return replaceOperand(SI, 1, Constant::getAllOnesValue(TrueVal->getType()));
```
Please add a test using i128 for this, which your current implementation miscompiles due to incorrect extension.

Reply:
I didn't even notice that, thanks for guiding me!